crates/uniplate/src/biplate.rs

```diff
@@ -0,0 +1,141 @@
+//#![cfg(feature = "unstable")]
+
+#![allow(clippy::type_complexity)]
+
+use std::sync::Arc;
+
+pub use super::Tree;
+use im;
+use im::vector;
+
+pub trait Biplate<To>
+where
+    Self: Sized + Clone + Eq + Uniplate + 'static,
+    To: Sized + Clone + Eq + Uniplate + 'static,
+{
+    /// Returns all the top most children of type to within from.
+    ///
+    /// If from == to then this function should return the root as the single child.
+    fn biplate(&self) -> (Tree<To>, Box<dyn Fn(Tree<To>) -> Self>);
+
+    /// Like descend but with more general types.
+    ///
+    /// If from == to then this function does not descend. Therefore, when writing definitions it
+    /// is highly unlikely that this function should be used in the recursive case. A common
+    /// pattern is to first match the types using descendBi, then continue the recursion with
+    /// descend.
+
+    fn descend_bi(&self, op: Arc<dyn Fn(To) -> To>) -> Self {
+        let (children, ctx) = self.biplate();
+        ctx(children.map(op))
+    }
+
+    // TODO (niklasdewally): Uniplate does some wierd Haskell I don't understand here.
+    // What are children and universe are meant to be?
```

Comment:
children gives a list of immediate children, universe gives all descendants of the given type. happy to discuss further.

Reply:
This is what i was referring to -
https://github.com/ndmitchell/uniplate/blob/66a2c55a7de0f5d8b0e437479719469244e00fa4/Data/Generics/Uniplate/Internal/OperationsInc.hs#L189

I have a correct implementation functionality wise, but the Haskell version doesn't use structure (what I call list) here at all, and does this wierd "builder" thing I do not understand...

Reply:
I've replaced this TODO with a comment.